### PR TITLE
When optimizerQuery is set then cache deserialized objects

### PR DESCRIPTION
**Pull request by jerrinot**
This part should be rewritten, this commit is just fixing buggy behaviour, not the problematic design.

### CODE REVIEW

**Comment by eminn** [comment]
I am not that familiar with the whole picture of the query mechanism, but this change looks good to merge :+1:

**Comment by ahmetmircik** [comment]
:+1:

**Comment by jerrinot** [comment]
verify

File: hazelcast/src/main/java/com/hazelcast/map/impl/record/Records.java [hunk 1]
```diff
@@ -63,4 +63,9 @@ public static Object getValueOrCachedValue(Record record, SerializationService s
         return value;
     }
 
+    private static boolean shouldCache(Record record, Object value) {
+        return (record instanceof CachedDataRecord || record instanceof CachedDataRecordWithStats)
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/map/impl/record/Records.java]: I would simplify this a bit. e.g.

```
if(!((record instanceof CachedDataRecord || record instanceof CachedDataRecordWithStats))
   return false;

if(....
```

The current implementation is painful to read.

**Comment by pveentjer** [comment]
Minor comment. :+1:

**Comment by jerrinot** [comment]
I refactored the `shouldCache()`  to make it more readable.

**Comment by jerrinot** [comment]
verify

**Pull request merged**